[release/6.0.1xx] Update BlazorWebAssembly targets to work with updated wasm targets

radical opened the pull request:
[This runtime PR](https://github.com/dotnet/runtime/pull/57556) adds support for
relinking the native wasm files during the build (instead of only during
publish).

This required changes to the targets available, and how they can be
used, and thus require changes in the blazor targets to work with that.

The new setup is explained in
https://github.com/radical/runtime/blob/wasm-build/src/mono/wasm/build/README.md
.

This will need to be merged *after* https://github.com/dotnet/runtime/pull/57556 goes in.

radical commented:
@javiercn @pranavkm @captainsafia The changes here might not be the best way to get the results, so please feel free to change that. But with this, I can build blazorwasm projects with native libraries, which should allow debugging with VS.

radical commented:
cc @lewing

javiercn commented:
@radical do you think we can have a meeting to discuss these changes?

I think we can probably find a more streamlined way to integrate these changes but I need a bit more context I think.

radical committed:
Update BlazorWebAssembly targets to work with updated wasm targets

[This PR](https://github.com/dotnet/runtime/pull/57556) adds support for
relinking the native wasm files during the build (instead of only during
publish).

This required changes to the targets available, and how they can be
used, and thus require changes in the blazor targets to work with that.

The new setup is explained in
https://github.com/radical/runtime/blob/wasm-build/src/mono/wasm/build/README.md .

radical committed:
Fix metadata name for skipping interop assembly for aot

javiercn commented on src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.6_0.targets:
```diff
@@ -406,7 +487,7 @@ Copyright (c) .NET Foundation. All rights reserved.
   </Target>
 
-  <Target Name="ProcessPublishFilesForBlazor" DependsOnTargets="_ResolveBlazorWasmConfiguration;LoadStaticWebAssetsBuildManifest" AfterTargets="ILLink">
+  <Target Name="ProcessPublishFilesForBlazor" DependsOnTargets="_ResolveBlazorWasmConfiguration;LoadStaticWebAssetsBuildManifest" AfterTargets="ILLink" Condition="'$(WasmBuildingForPublish)' != 'true'">
```
I'm not 100% sure about the condition here.

This target computes the list of publish assets for the app, so it should always run.

radical committed:
Merge remote-tracking branch 'origin/release/6.0.1xx' into release/6.0.1xx

radical committed:
Update BlazorWebAssembly.6_0.targets

1. Ensure that the PublishDir uses the same case as PublishUrl, to
   workaround a bug where we end up with two directories on linux:
   `Publish`, and `publish`.
   https://github.com/dotnet/sdk/issues/20923

2. Add all the `@(WasmNativeAssets)` as candidates, and not just
   `.wasm`

3. And update to track property name change

javiercn commented on src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.6_0.targets:
```diff
@@ -115,12 +124,75 @@ Copyright (c) .NET Foundation. All rights reserved.
+        <CopyLocal>true</CopyLocal>
+      </_DotNetJsItem>
+    </ItemGroup>
+
```
Based on offline discussion, we don't need the scrambling here since @TanayParikh took care of it on his PR

radical replied:
Does that mean that the two targets `_Scramble*` can be dropped completely?

radical replied:
Ah, the target is no longer there! 👍

javiercn replied:
Yep. We do this all the time now on build and publish if necessary, as it improves the cache busting

javiercn reviewed (approved):
Overall changes look good.

I'm signing off to make sure things are unblocked (provided tests pass and trusting manual verification of the AoT scenarios has happened before merging.

radical committed:
blazorwasm: Update to track recent changes for publish

Thanks to @javiercn

radical committed:
Merge remote-tracking branch 'origin/release/6.0.1xx' into blazor-build-publish-6.0.1

radical commented:
The corresponding runtime PR https://github.com/dotnet/runtime/pull/58913 has merged. Next is to wait for a version bump PR to be opened.

TanayParikh commented:
I believe this'll need to be cherry-picked into `main` as well.
